### ISSUE TITLE
Array of pairs don't do proper type checks if they are a part of dictionary

### ISSUE THREAD

**Issue by zakimaksyutov** [microsoft/TypeScript#26335]
**TypeScript Version:**  2.9.2 (from VS Code)

**Search Terms:**

**Code**

```
export class ArrayWrapper<T> {
    values: Array<[number, T]>;
}

// correctly reports error: Type 'number' is not assignable to type 'boolean'
export var instance: ArrayWrapper<boolean> = {
    values: [[1, false], [0, 10]]
};

// doesn't catch error: the second element of an array has "10" instead of boolean
export var instances: {[key: string]: ArrayWrapper<any>} = {
    instance: <ArrayWrapper<boolean>> {
        values: [[1, false], [0, 10]]
    }
};

// correctly reports error: Type 'number' is not comparable to type 'boolean'
export var instances2: {[key: string]: ArrayWrapper<any>} = {
    instance: <ArrayWrapper<boolean>> {
        values: [[1, 11], [0, 10]]
    }
};
```

**Expected behavior:** _values: [[1, false], [0, **10**]]_ should be flagged as error since 10 is a number and not a boolean

**Actual behavior:** compiler doesn't catch type violation

**Comment by andy-ms** [comment]
`<T> {}` is a type assertion -- it's unsafe and allows you to silence compiler errors.
For a safer version you might do something like:
```ts
function as<T>(x: T): T { return x; }
as<ArrayWrapper<boolean>>({
    values: [[1, false], [0, 10]], // Correctly get error
})
```
See also #7481.

**Comment by andy-ms** [comment]
Type assertions have a heuristic for whether a type looks like it might possibly be assignable to another. If at least *one* of the array elements has a boolean element, the array is `(number | boolean)[][]` which looks like it's possible a `[number, boolean][]`. But without that, it's a `number[][]` which doesn't look like it's possibly a `[number, boolean][]`.
